include/deal.II/matrix_free/shape_info.templates.h

```diff
@@ -186,122 +192,167 @@ namespace internal
-                      }
-                }
+              // check if we are a Hermite type
+              element_type = tensor_symmetric_hermite;
+              for (unsigned int i=1; i<n_dofs_1d; ++i)
+                if (std::abs(this->shape_data_on_face[0][i][0]) > 1e-12)
+                  element_type = tensor_symmetric;
+              for (unsigned int i=2; i<n_dofs_1d; ++i)
+                if (std::abs(this->shape_data_on_face[0][n_dofs_1d+i][0]) > 1e-12)
+                  element_type = tensor_symmetric;
             }
         }
       else if (element_type == tensor_symmetric_plus_dg0)
         check_1d_shapes_symmetric(n_q_points_1d);
 
-      // face information
-      unsigned int n_faces = GeometryInfo<dim>::faces_per_cell;
-      this->face_indices.reinit(n_faces, this->dofs_per_face);
-      switch (dim)
-        {
-        case 3:
-        {
-          for (unsigned int i=0; i<this->dofs_per_face; i++)
-            {
-              const unsigned int jump_term =
-                this->dofs_per_face*((i*n_dofs_1d)/this->dofs_per_face);
-              this->face_indices(0,i) = i*n_dofs_1d;
-              this->face_indices(1,i) = i*n_dofs_1d + n_dofs_1d-1;
-              this->face_indices(2,i) = i%n_dofs_1d + jump_term;
-              this->face_indices(3,i) = (i%n_dofs_1d + jump_term +
-                                         (n_dofs_1d-1)*n_dofs_1d);
-              this->face_indices(4,i) = i;
-              this->face_indices(5,i) = (n_dofs_1d-1)*this->dofs_per_face+i;
-            }
-          break;
-        }
-        case 2:
+      nodal_at_cell_boundaries = true;
+      for (unsigned int i=1; i<n_dofs_1d; ++i)
+        if (std::abs(this->shape_data_on_face[0][i][0]) > 1e-13 ||
+            std::abs(this->shape_data_on_face[1][i-1][0]) > 1e-13)
+          nodal_at_cell_boundaries = false;
+
+      if (nodal_at_cell_boundaries == true)
         {
-          for (unsigned int i=0; i<this->dofs_per_face; i++)
+          face_to_cell_index_nodal.reinit(GeometryInfo<dim>::faces_per_cell,
+                                          dofs_per_face);
+          for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
             {
-              this->face_indices(0,i) = n_dofs_1d*i;
-              this->face_indices(1,i) = n_dofs_1d*i + n_dofs_1d-1;
-              this->face_indices(2,i) = i;
-              this->face_indices(3,i) = (n_dofs_1d-1)*n_dofs_1d+i;
+              const unsigned int direction = f/2;
+              const unsigned int stride = direction < dim-1 ? (fe_degree+1) : 1;
+              int shift = 1;
+              for (unsigned int d=0; d<direction; ++d)
+                shift *= fe_degree+1;
+              const unsigned int offset = (f%2)*fe_degree*shift;
+
+              if (direction == 0 || direction == dim-1)
+                for (unsigned int i=0; i<dofs_per_face; ++i)
+                  face_to_cell_index_nodal(f,i) = offset + i*stride;
```

Comment:
Can you add a comment on the mapping here, I am not sure what `offset + i*stride` represents.